cmd/sourced/compose/workdir/factory.go

```diff
@@ -224,31 +276,96 @@ func (f *envFile) UnmarshalEnv(b []byte) error {
 		}
 
 		parts := strings.SplitN(line, "=", 2)
+		name := parts[0]
 		value := parts[1]
-
-		switch parts[0] {
-		case "COMPOSE_PROJECT_NAME":
-			f.Workdir = strings.TrimPrefix(value, "srcd-")
-		case "GITBASE_VOLUME_SOURCE":
-			f.ReposDir = value
-		case "GITHUB_ORGANIZATIONS":
-			if value != "" {
-				f.GithubOrganizations = strings.Split(value, ",")
+		field := v.FieldByName(varToName(name))
+		// skip unknown values
+		if !field.IsValid() {
+			continue
+		}
+		// skip empty values
+		if value == "" {
+			continue
+		}
+		switch field.Kind() {
+		case reflect.String:
+			field.SetString(value)
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			i, err := strconv.ParseInt(value, 10, 64)
+			if err != nil {
+				return fmt.Errorf("can't parse variable %s with value %s: %v", name, value, err)
+			}
+			field.SetInt(i)
+		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+			i, err := strconv.ParseUint(value, 10, 64)
+			if err != nil {
+				return fmt.Errorf("can't parse variable %s with value %s: %v", name, value, err)
+			}
+			field.SetUint(i)
+		case reflect.Float32, reflect.Float64:
+			i, err := strconv.ParseFloat(value, 64)
+			if err != nil {
+				return fmt.Errorf("can't parse variable %s with value %s: %v", name, value, err)
+			}
+			field.SetFloat(i)
+		case reflect.Bool:
+			if value == "true" {
+				field.SetBool(true)
+			} else {
+				field.SetBool(false)
 			}
-		case "GITHUB_TOKEN":
-			f.GithubToken = value
-		case "NO_FORKS":
-			if value == "false" {
-				f.WithForks = true
+		case reflect.Slice:
+			if field.Type().Elem().Kind() != reflect.String {
+				panic("only slices of strings are supported")
 			}
+			vs := strings.Split(value, ",")
+			slice := reflect.MakeSlice(field.Type(), len(vs), len(vs))
+			for i, v := range vs {
+				slice.Index(i).SetString(v)
+			}
+			field.Set(slice)
+		default:
+			panic(fmt.Sprintf("unsupported type: %v", field.Kind()))
+		}
+	}
+
+	return scanner.Err()
+}
+
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+func nameToVar(name string) string {
+	name = matchAllCap.ReplaceAllString(name, "${1}_${2}")
+	return strings.ToUpper(name)
+}
+
+var nameExceptions = map[string]bool{
```

Comment:
If we need name exceptions, and the process is not fully automatic, I wonder if the code would be easier to understand using tags, e.g.:
```go
type envFile struct {
	ComposeProjectName string `env:"COMPOSE_PROJECT_NAME"`

	GitbaseVolumeType   string `env:"GITBASE_VOLUME_TYPE"`
...
}
```

Reply:
I wish we could just import https://github.com/golang/lint/blob/master/lint.go#L770 but it's private. 

I'll replace this code with https://github.com/serenize/snaker so it will work the same as https://github.com/src-d/envconfig which we are using already. There is no problem to add support for tags but we don't really need it right now and I'm not sure we will ever need.

Reply:
I mean here is already kind of unused method `Unmarshal`. But at least I see how it can be used. Let's avoid adding "dead" code.